src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -297,28 +324,8 @@ void grpc_tcp_client_create_from_prepared_fd(
 
   grpc_pollset_set_add_fd(interested_parties, fdobj);
 
-  async_connect* ac = new async_connect();
-  ac->closure = closure;
-  ac->ep = ep;
-  ac->fd = fdobj;
-  ac->interested_parties = interested_parties;
-  ac->addr_str = grpc_sockaddr_to_uri(addr);
-  gpr_mu_init(&ac->mu);
-  ac->refs = 2;
-  GRPC_CLOSURE_INIT(&ac->write_closure, on_writable, ac,
-                    grpc_schedule_on_exec_ctx);
-  ac->channel_args = grpc_channel_args_copy(channel_args);
-
-  if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
-    gpr_log(GPR_INFO, "CLIENT_CONNECT: %s: asynchronously connecting fd %p",
-            ac->addr_str.c_str(), fdobj);
-  }
-
-  gpr_mu_lock(&ac->mu);
-  GRPC_CLOSURE_INIT(&ac->on_alarm, tc_on_alarm, ac, grpc_schedule_on_exec_ctx);
-  grpc_timer_init(&ac->alarm, deadline, &ac->on_alarm);
-  grpc_fd_notify_on_write(ac->fd, &ac->write_closure);
-  gpr_mu_unlock(&ac->mu);
+  new AsyncConnect(fdobj, interested_parties, addr, ep, closure, channel_args,
```

Comment:
This is an odd API from the caller's perspective, it works entirely by side effect. It would be nice if the usage conveyed that the object needs no owner and is self-deleting. A RefCounted type is probably the answer here given that it has a `refs_` member. And consider maybe changing the class name or adding a factory method to convey it's self-deleting nature.